File: sdks/dotnet/src/Backlex/Collection.cs
namespace Backlex;

/// <summary>
/// A typed CRUD handle for one collection. Obtain via <c>client.From&lt;T&gt;(slug)</c>.
/// Use <c>Dictionary&lt;string, object?&gt;</c> as T for schema-blind access, or a
/// generated/POCO type.
/// </summary>
public sealed class Collection<T>
{
    private readonly BacklexClient _client;
    private readonly string _slug;

    internal Collection(BacklexClient client, string slug)
    {
        _client = client;
        _slug = slug;
    }

    public async Task<ListResponse<T>> ListAsync(ListQuery? query = null) =>
        await _client.RequestAsync<ListResponse<T>>(
            HttpMethod.Get, $"/api/items/{_slug}{BacklexClient.BuildSearch(query)}")
            .ConfigureAwait(false) ?? new ListResponse<T>();

    /// <summary>Fluent builder that compiles to a <see cref="ListQuery"/>.</summary>
    public QueryBuilder<T> Query() => new(ListAsync);

    public async Task<ItemResponse<T>> OneAsync(string id) =>
        await _client.RequestAsync<ItemResponse<T>>(HttpMethod.Get, $"/api/items/{_slug}/{id}")
            .ConfigureAwait(false) ?? new ItemResponse<T>();

    public async Task<ItemResponse<T>> CreateAsync(object data) =>
        await _client.RequestAsync<ItemResponse<T>>(HttpMethod.Post, $"/api/items/{_slug}", data)
            .ConfigureAwait(false) ?? new ItemResponse<T>();

    public async Task<ItemResponse<T>> UpdateAsync(string id, object patch) =>
        await _client.RequestAsync<ItemResponse<T>>(HttpMethod.Patch, $"/api/items/{_slug}/{id}", patch)
            .ConfigureAwait(false) ?? new ItemResponse<T>();

    public async Task<DeleteResult> DeleteAsync(string id) =>
        await _client.RequestAsync<DeleteResult>(HttpMethod.Delete, $"/api/items/{_slug}/{id}")
            .ConfigureAwait(false) ?? new DeleteResult();
}

File: sdks/ruby/lib/backlex/collection.rb
# frozen_string_literal: true

module Backlex
  # A CRUD handle for one collection. Obtain via client.from("slug").
  class Collection
    def initialize(client, slug)
      @client = client
      @slug = slug
    end

    def list(query = nil)
      @client.request("GET", "/api/items/#{@slug}#{Client.build_search(query)}")
    end

    # Fluent builder that compiles to a ListQuery.
    def query
      QueryBuilder.new(method(:list))
    end

    # Single-function aggregate (count/sum/avg/min/max), optionally grouped.
    # body = { "agg" => "sum", "field" => "price", "groupBy" => "status" }
    def aggregate(body)
      @client.request("POST", "/api/items/#{@slug}/aggregate", body)
    end

    # query may carry expand/locale, the same params the list endpoint accepts.
    def one(id, query = nil)
      @client.request("GET", "/api/items/#{@slug}/#{id}#{Client.build_search(query)}")
    end

    def create(data)
      @client.request("POST", "/api/items/#{@slug}", data)
    end

    def update(id, patch)
      @client.request("PATCH", "/api/items/#{@slug}/#{id}", patch)
    end

    def delete(id)
      @client.request("DELETE", "/api/items/#{@slug}/#{id}")
    end

    # Flip a versioned item to published.
    def publish(id)
      @client.request("POST", "/api/items/#{@slug}/#{id}/publish")
    end

    # Flip a versioned item back to draft.
    def unpublish(id)
      @client.request("POST", "/api/items/#{@slug}/#{id}/publish?unpublish=1")
    end
  end
end
